src/main/common/filter.c

```diff
@@ -68,15 +85,26 @@ FAST_CODE float pt1FilterApply(pt1Filter_t *filter, float input)
 
 // PT2 Low Pass filter
 
+// PTn cutoff correction = 1 / sqrt(2^(1/n) - 1)
+#define CUTOFF_CORRECTION_PT2 1.553773974f
+
 FAST_CODE float pt2FilterGain(float f_cut, float dT)
 {
-    // PTn cutoff correction = 1 / sqrt(2^(1/n) - 1)
-    #define CUTOFF_CORRECTION_PT2 1.553773974f
-
     // shift f_cut to satisfy -3dB cutoff condition
     return pt1FilterGain(f_cut * CUTOFF_CORRECTION_PT2, dT);
 }
 
+// Calculates filter gain based on delay (time constant of filter) - time it takes for filter response to reach 63.2% of a step input.
+FAST_CODE float pt2FilterGainFromDelay(float delay, float dT)
+{
+    if (delay == 0) {
+        return 1.0f; // gain = 1 means no filtering
+    }
+
+    const float cutoffHz = 1.0f / (M_PIf * delay * CUTOFF_CORRECTION_PT2);
```

Comment:
Have you confirmed that this will actually give you the time constant that is equal to that of a PT1 filter? I'm not so sure about that. I need to look into this first.

Reply:
@KarateBrot I did this simple math:
PT2 at 100hz is two of PT1(100hz * CORRECTION) applied in series in BF code.
So the time constant (delay) is a sum of these two pt1 delays with corrected cutoffs.
"Delay" for a single PT1 in this case is `1 / (2 * Pi * 100hz * correction)`.
Delay of initial PT2 is two times more, so `1 / (Pi * 100hz * correction)`.
Or if we solving Cutoff frequency from delay:
`Cutoff_hz = 1 / (Pi * delay * correction)`

Reply:
This equation with the cutoff correction is an approximation to get very similar cutoff frequencies for PT1, PT2 and PT3 filters. But the same cutoff frequency results in very different time constants (if you define the time constant as the time it takes the filter to rise to 63.3% of its final value).

I guess it comes down to how you define the time constant. Because if you want the time constant to have the same meaning for PT1, PT2, PT3, etc... then this doesn't work for anything other than PT1. As you can see in this picture, the cutoffs are approximately the same but the time constants are completely different in the step response plot.

![grafik](https://github.com/betaflight/betaflight/assets/19867640/4f4f55f3-06d8-4de3-ad27-03847762a634)
![grafik](https://github.com/betaflight/betaflight/assets/19867640/46a3c0c4-ee6b-47b5-a40f-bd1b9d855798)

Reply:
@KarateBrot if PT1 and PT2 have same cutoff values, of course the time constant will be different for them, and that's reflected in the functions i made.

What cutoff values were used in these plots for PT1, PT2 and PT3?
And could you also plz plot the phase delay? Phase delay converges to time constant when frequency approaches 0.

Reply:
I guess, if we say it is the time constant of the staggered PT1s that make up PT2, but not the time constant of the PT2, then it makes sense. But then we don't need the functions in `filter.c` and only need a simple conversion where you initialize the PT2 filter.

Reply:
The cutoff was 100Hz but it could be anything, as long as it's the same for all filters to compare them.

The delay plots I gave you are dependent on frequency. The delay for each frequency component in a signal is different. It's not a single "delay for the whole signal".

Reply:
> Phase delay converges to time constant when frequency approaches 0.

So you want to only look at very low frequecies?

Reply:
so for 100hz PT1 the time constant will be:
`1/(2*PI * 100) = 0.00159`

for 100hz PT2 the time constant will be:
`1/(Pi * 100 * 1.553773974) = 0.00205`

Both ~match with your step response plot.

Reply:
But yeah, basically "delay" here is just how fast TPA reacts on throttle moves when throttle is being moved by a human. Basically a time constant.

Reply:
Maybe I am looking at it the wrong way and my brain is playing tricks on me 🤔 Fine by me then. I'll do a new review this evening 👍

Reply:
@KarateBrot : The trick is to adjust cutoff of higher order PTn to get expected delay. Delay of PT1 cascade is about linear with number of stages, so you just divide total delay by number of stages and undo CUTOFF_CORRECTION_PTn that is applied inside filter coefficient calculation.

You can also divide total delay by number of stages and use this partial adelay to calculate cutoff of (each) pt1 stage. Actual attenuation at this (1-stage) frequency is n*3db (CUTOFF_CORRECTION_PT2 adjusts cutoff so that total attenuation is 3db)